core/server/master/src/main/java/alluxio/master/file/contexts/OperationContext.java

```diff
@@ -11,54 +11,89 @@
 
 package alluxio.master.file.contexts;
 
+import com.google.protobuf.GeneratedMessageV3;
+
 import javax.annotation.concurrent.NotThreadSafe;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
 
 /**
  * Used as a base class for wrapping context around proto messages.
  *
  * @param <T> Proto message type
+ * @param <C> extended type
  */
 @NotThreadSafe
-public class OperationContext<T extends com.google.protobuf.GeneratedMessageV3.Builder<?>> {
+public class OperationContext<T extends GeneratedMessageV3.Builder, C extends OperationContext> {
   // Proto message that is being wrapped
   private T mOptionsBuilder;
   // Used to track client call status.
-  private CallTracker mCallTracker;
+  private List<CallTracker> mCallTrackers;
 
   /**
    * Creates an instance with given proto message.
    *
    * @param optionsBuilder Internal proto message builder instance
    */
   public OperationContext(T optionsBuilder) {
-    this(optionsBuilder, null);
     mOptionsBuilder = optionsBuilder;
-    mCallTracker = CallTracker.NOOP_TRACKER;
+    mCallTrackers = new LinkedList<>();
   }
 
   /**
-   * Creates an instance with given proto message.
+   * @return underlying proto message instance
+   */
+  public T getOptions() {
+    return mOptionsBuilder;
+  }
+
+  /**
+   * Used to transfer trackers of a context to this instance.
+   * This is required when creating internal contexts.
    *
-   * @param optionsBuilder Internal proto message builder instance
-   * @param callTracker client call tracker, or {@code null} if no tracking is desired
+   * @param context the source context
+   * @return the updated instance
    */
-  public OperationContext(T optionsBuilder, CallTracker callTracker) {
-    mOptionsBuilder = optionsBuilder;
-    mCallTracker = callTracker;
+  public C withTracker(OperationContext context) {
```

Comment:
I think it is strange to have an API called `withTracker` but the input parameter is not a tracker but a context? Why can't we just use the other implementation only?

Reply:
I still have this question.

Reply:
turned out it was a dead method. removed.